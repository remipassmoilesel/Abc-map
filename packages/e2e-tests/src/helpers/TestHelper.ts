/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import Chainable = Cypress.Chainable;
import { Routes } from './Routes';

export class TestHelper {
  public static init(): Chainable<any> {
    return (
      cy
        // We set viewport
        .viewport(1980, 1080)
        // We set default language
        .visit(Routes.map().format(), {
          onBeforeLoad(win) {
            Object.defineProperty(win.navigator, 'language', { value: 'en-US' });
            Object.defineProperty(win.navigator, 'languages', { value: ['en'] });
            Object.defineProperty(win.navigator, 'accept_languages', { value: ['en'] });
          },
        })
        // We clear local storage data (and redux store persistence)
        // Despite docs says it run automatically before each test, it does not clear side menu states
        .clearLocalStorage()
    );
  }
}
