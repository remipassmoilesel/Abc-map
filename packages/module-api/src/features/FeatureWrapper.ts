/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import Feature from 'ol/Feature';
import Geometry from 'ol/geom/Geometry';
import { Circle, GeometryCollection, LinearRing, LineString, MultiLineString, MultiPoint, MultiPolygon, Point, Polygon } from 'ol/geom';
import { FeatureStyle } from './FeatureStyle';

export declare type OlGeometry =
  | Geometry
  | Point
  | LineString
  | LinearRing
  | Polygon
  | MultiPoint
  | MultiLineString
  | MultiPolygon
  | GeometryCollection
  | Circle;

export declare type PropertiesMap = { [key: string]: any };
export declare type DataPropertiesMap = { [key: string]: string | boolean | number | null | undefined };

/**
 * This class is used to complete OpenLayers features (selection, styling, serialisation, ...)
 *
 * The goal is not to replace the use of Openlayers features, but to extend them by composition.
 *
 * Call unwrap() to access the underlying feature.
 */
export interface FeatureWrapper<Geom extends OlGeometry = OlGeometry> {
  unwrap(): Feature<Geom>;
  setId(id?: string | number): FeatureWrapper;
  getId(): string | number | undefined;
  clone(): FeatureWrapper;
  isSelected(): boolean;
  setSelected(value: boolean): FeatureWrapper;
  getStyleProperties(): FeatureStyle;
  setStyleProperties(properties: FeatureStyle): FeatureWrapper;
  setDefaultStyle(): FeatureWrapper;
  getGeometry(): Geom | undefined;
  hasGeometry(...geoms: string[]): boolean;
  setText(text: string): FeatureWrapper;
  getText(): string | undefined;
  getAllProperties(): PropertiesMap;
  getDataProperties(): DataPropertiesMap;
  setProperties(properties: DataPropertiesMap): FeatureWrapper;
  setDataProperties(properties: DataPropertiesMap): FeatureWrapper;
}

/* eslint-disable @typescript-eslint/no-unused-vars */

export class DumbFeatureWrapper implements FeatureWrapper {
  public clone(): FeatureWrapper {
    return {} as any;
  }

  public getAllProperties(): PropertiesMap {
    return {} as any;
  }

  public getGeometry(): OlGeometry | undefined {
    return {} as any;
  }

  public getId(): string | number | undefined {
    return {} as any;
  }

  public getDataProperties(): DataPropertiesMap {
    return {} as any;
  }

  public getStyleProperties(): FeatureStyle {
    return {} as any;
  }

  public getText(): string | undefined {
    return {} as any;
  }

  public hasGeometry(...geoms: string[]): boolean {
    return {} as any;
  }

  public isSelected(): boolean {
    return {} as any;
  }

  public setDataProperties(properties: DataPropertiesMap): FeatureWrapper {
    return {} as any;
  }

  public setDefaultStyle(): FeatureWrapper {
    return {} as any;
  }

  public setId(id?: string | number): FeatureWrapper {
    return {} as any;
  }

  public setProperties(properties: DataPropertiesMap): FeatureWrapper {
    return {} as any;
  }

  public setSelected(value: boolean): FeatureWrapper {
    return {} as any;
  }

  public setStyleProperties(properties: FeatureStyle): FeatureWrapper {
    return {} as any;
  }

  public setText(text: string): FeatureWrapper {
    return {} as any;
  }

  public unwrap(): Feature<OlGeometry> {
    return {} as any;
  }
}
