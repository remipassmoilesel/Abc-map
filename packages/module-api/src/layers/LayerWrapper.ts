/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import TileLayer from 'ol/layer/Tile';
import TileSource from 'ol/source/Tile';
import VectorSource from 'ol/source/Vector';
import Geometry from 'ol/geom/Geometry';
import VectorImageLayer from 'ol/layer/VectorImage';
import { Source, TileWMS, WMTS, XYZ } from 'ol/source';
import { Layer } from 'ol/layer';
import LayerRenderer from 'ol/renderer/Layer';
import { AbcProjection } from '../map/AbcProjection';

export type OlLayers = Layer<Source, LayerRenderer<any>> | VectorImageLayer<VectorSource<Geometry>> | TileLayer<TileSource>;
export type OlSources = Source | VectorSource<Geometry> | TileSource | TileWMS | WMTS;

export type VectorLayerWrapper = LayerWrapper<VectorImageLayer<VectorSource<Geometry>>, VectorSource<Geometry>>;
export type PredefinedLayerWrapper = LayerWrapper<TileLayer<TileSource>, TileSource>;
export type WmsLayerWrapper = LayerWrapper<TileLayer<TileSource>, TileWMS>;
export type WmtsLayerWrapper = LayerWrapper<TileLayer<TileSource>, WMTS>;
export type XyzLayerWrapper = LayerWrapper<TileLayer<TileSource>, XYZ>;

export type Meta = { [k: string]: any };

/**
 * LayerWrapper wraps Openlayers layers, mainly for Abc-Map specific critical operations (selection, naming, ...)
 *
 * The goal is not to replace the use of Openlayers layers, but to extend them by composition.
 *
 * Call unwrap() to access the underlying layer.
 */
export interface LayerWrapper<Layer extends OlLayers = OlLayers, Source extends OlSources = OlSources> {
  unwrap(): Layer;
  getSource(): Source;
  setId(value?: string): LayerWrapper;
  getId(): string | undefined;
  setName(value: string): LayerWrapper;
  getName(): string | undefined;
  setActive(value: boolean): LayerWrapper;
  isActive(): boolean;
  setVisible(value: boolean): LayerWrapper;
  isVisible(): boolean;
  setOpacity(value: number): LayerWrapper;
  getOpacity(): number;
  getType(): LayerType | undefined;
  isPredefined(): this is PredefinedLayerWrapper;
  isVector(): this is VectorLayerWrapper;
  isWms(): this is WmsLayerWrapper;
  isWmts(): this is WmtsLayerWrapper;
  isXyz(): this is XyzLayerWrapper;
  getAttributions(format: AttributionFormat): string[] | undefined;
  setAttributions(attr: string[]): LayerWrapper<Layer, Source>;
  getProjection(): AbcProjection | undefined;
}

export enum LayerType {
  Vector = 'Vector',
  Predefined = 'Predefined',
  Wms = 'Wms',
  Wmts = 'Wmts',
  Xyz = 'Xyz',
}

export enum AttributionFormat {
  Text = 'Text',
  HTML = 'HTML',
}
