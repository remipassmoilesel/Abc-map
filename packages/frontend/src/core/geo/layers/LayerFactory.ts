/**
 * Copyright © 2021 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import TileLayer from 'ol/layer/Tile';
import { OSM, Stamen, TileWMS, WMTS, XYZ } from 'ol/source';
import uuid from 'uuid-random';
import VectorSource from 'ol/source/Vector';
import { tileLoadingAuthenticated } from './tileLoadingAuthenticated';
import { AbcProjection, LayerType, PredefinedLayerModel, PredefinedMetadata, VectorMetadata, WmsMetadata, WmtsMetadata, XyzMetadata } from '@abc-map/shared';
import { LayerWrapper, PredefinedLayerWrapper, VectorLayerWrapper, WmsLayerWrapper, WmtsLayerWrapper, XyzLayerWrapper } from './LayerWrapper';
import VectorImageLayer from 'ol/layer/VectorImage';
import Geometry from 'ol/geom/Geometry';
import TileSource from 'ol/source/Tile';
import { styleFunction } from '../styles/style-function';
import { tileLoadingPublic } from './tileLoadingPublic';
import { WmsSettings, WmtsSettings } from './LayerFactory.types';
import { prefixedTranslation } from '../../../i18n/i18n';
import { DefaultStyleOptions } from '../styles/StyleFactoryOptions';

const t = prefixedTranslation('LayerFactory:');

export class LayerFactory {
  public static newPredefinedLayer(model: PredefinedLayerModel): PredefinedLayerWrapper {
    let name: string;
    let source: TileSource;
    switch (model) {
      case PredefinedLayerModel.OSM:
        source = new OSM();
        name = t('OpenStreetMap');
        break;
      case PredefinedLayerModel.StamenToner:
        source = new Stamen({ layer: 'toner' });
        name = t('Stamen_Toner');
        break;
      case PredefinedLayerModel.StamenTonerLite:
        source = new Stamen({ layer: 'toner-lite' });
        name = t('Stamen_Toner_Lite');
        break;
      case PredefinedLayerModel.StamenTerrain:
        source = new Stamen({ layer: 'terrain' });
        name = t('Stamen_Terrain');
        break;
      case PredefinedLayerModel.StamenWatercolor:
        source = new Stamen({ layer: 'watercolor' });
        name = t('Stamen_Watercolor');
        break;
      default:
        throw new Error(`Unhandled predefined layer type: ${model}`);
    }

    const layer = new TileLayer({ source });

    const metadata: PredefinedMetadata = {
      id: uuid(),
      name,
      type: LayerType.Predefined,
      active: false,
      opacity: 1,
      visible: true,
      model,
    };

    return LayerWrapper.from<TileLayer<TileSource>, TileSource, PredefinedMetadata>(layer).setMetadata(metadata);
  }

  public static newVectorLayer(source?: VectorSource<Geometry>): VectorLayerWrapper {
    const _source = source || new VectorSource();
    const layer = new VectorImageLayer({ style: (f) => styleFunction(DefaultStyleOptions, f), source: _source });

    const metadata: VectorMetadata = {
      id: uuid(),
      name: t('Geometries'),
      type: LayerType.Vector,
      active: false,
      opacity: 1,
      visible: true,
    };

    return LayerWrapper.from<VectorImageLayer<VectorSource<Geometry>>, VectorSource<Geometry>, VectorMetadata>(layer).setMetadata(metadata);
  }

  public static newWmsLayer(settings: WmsSettings): WmsLayerWrapper {
    const tileLoadFunction = settings.auth?.username && settings.auth?.password ? tileLoadingAuthenticated(settings.auth) : tileLoadingPublic();

    const layer = new TileLayer({
      extent: settings.extent,
      source: new TileWMS({
        urls: settings.remoteUrls,
        params: { LAYERS: settings.remoteLayerName, TILED: true },
        tileLoadFunction,
        projection: settings.projection?.name,
      }),
    });

    const metadata: WmsMetadata = {
      id: uuid(),
      name: settings.remoteLayerName,
      type: LayerType.Wms,
      active: false,
      opacity: 1,
      visible: true,
      remoteUrls: settings.remoteUrls,
      remoteLayerName: settings.remoteLayerName,
      projection: settings.projection,
      extent: settings.extent,
      auth: settings.auth,
    };

    return LayerWrapper.from<TileLayer<TileSource>, TileWMS, WmsMetadata>(layer).setMetadata(metadata);
  }

  public static newWmtsLayer(settings: WmtsSettings): WmtsLayerWrapper {
    if (!settings.sourceOptions) {
      throw new Error('Source options are mandatory');
    }

    const tileLoadFunction = settings.auth?.username && settings.auth?.password ? tileLoadingAuthenticated(settings.auth) : tileLoadingPublic();

    // We do not set extent here as it can have weird behaviors
    const layer = new TileLayer({
      source: new WMTS({
        ...settings.sourceOptions,
        tileLoadFunction,
      }),
    });

    const metadata: WmtsMetadata = {
      id: uuid(),
      name: settings.remoteLayerName,
      type: LayerType.Wmts,
      active: false,
      opacity: 1,
      visible: true,
      capabilitiesUrl: settings.capabilitiesUrl,
      remoteLayerName: settings.remoteLayerName,
      auth: settings.auth,
    };

    return LayerWrapper.from<TileLayer<TileSource>, WMTS, WmtsMetadata>(layer).setMetadata(metadata);
  }

  public static newXyzLayer(url: string, projection?: AbcProjection): XyzLayerWrapper {
    const source = new XYZ({ url, projection: projection?.name, tileLoadFunction: tileLoadingPublic() });
    const layer = new TileLayer({ source });

    const metadata: XyzMetadata = {
      id: uuid(),
      name: t('XYZ_layer'),
      type: LayerType.Xyz,
      active: false,
      opacity: 1,
      visible: true,
      remoteUrl: url,
      projection,
    };

    return LayerWrapper.from<TileLayer<TileSource>, XYZ, XyzMetadata>(layer).setMetadata(metadata);
  }
}
