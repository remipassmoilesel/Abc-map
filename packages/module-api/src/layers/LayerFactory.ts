/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { PredefinedLayerWrapper, VectorLayerWrapper, WmsLayerWrapper, WmtsLayerWrapper, XyzLayerWrapper } from './LayerWrapper';
import Geometry from 'ol/geom/Geometry';
import VectorSource from 'ol/source/Vector';
import { PredefinedLayerModel } from './PredefinedLayerModel';
import { AbcProjection } from '../map';
import { Options as WmtsSourceOptions } from 'ol/source/WMTS';
import { BasicAuthentication } from '../utils';

export type LayerFactory = {
  newPredefinedLayer(model: PredefinedLayerModel): PredefinedLayerWrapper;
  newVectorLayer(source?: VectorSource<Geometry>): VectorLayerWrapper;
  newWmsLayer(settings: WmsSettings): WmsLayerWrapper;
  newWmtsLayer(settings: WmtsSettings): WmtsLayerWrapper;
  newXyzLayer(url: string, projection?: AbcProjection): XyzLayerWrapper;
};

export interface WmsSettings {
  capabilitiesUrl?: string;
  remoteUrls: string[];
  remoteLayerName: string;
  projection?: AbcProjection;
  extent?: [number, number, number, number];
  auth?: BasicAuthentication;
}

export interface WmtsSettings {
  capabilitiesUrl: string;
  remoteLayerName: string;
  sourceOptions?: WmtsSourceOptions;
  auth?: BasicAuthentication;
}
