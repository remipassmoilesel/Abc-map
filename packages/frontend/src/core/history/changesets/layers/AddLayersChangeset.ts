/**
 * Copyright © 2021 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { Changeset } from '../../Changeset';
import { MapWrapper } from '../../../geo/map/MapWrapper';
import { LayerWrapper } from '../../../geo/layers/LayerWrapper';
import { getServices } from '../../../Services';

export class AddLayersChangeset extends Changeset {
  public static create(layers: LayerWrapper[]): AddLayersChangeset {
    const { geo } = getServices();
    const map = geo.getMainMap();
    return new AddLayersChangeset(map, layers);
  }

  constructor(private map: MapWrapper, private layers: LayerWrapper[]) {
    super();
  }

  public async execute(): Promise<void> {
    this.layers.forEach((lay) => this.map.addLayer(lay));

    // We activate the last layer
    const layers = this.map.getLayers();
    if (layers.length) {
      this.map.setActiveLayer(layers[layers.length - 1]);
    }
  }

  public async undo(): Promise<void> {
    this.layers.forEach((lay) => this.map.removeLayer(lay));

    // We activate the last layer
    const layers = this.map.getLayers();
    if (layers.length) {
      this.map.setActiveLayer(layers[layers.length - 1]);
    }
  }
}
