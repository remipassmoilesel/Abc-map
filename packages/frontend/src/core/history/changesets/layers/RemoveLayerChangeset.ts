/**
 * Copyright © 2021 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { Changeset } from '../../Changeset';
import { MapWrapper } from '../../../geo/map/MapWrapper';
import { LayerWrapper } from '../../../geo/layers/LayerWrapper';
import { getServices } from '../../../Services';

export class RemoveLayerChangeset extends Changeset {
  public static create(layer: LayerWrapper) {
    const { geo } = getServices();
    const map = geo.getMainMap();

    const index = map.getLayers().findIndex((lay) => lay.getId() === layer.getId());
    return new RemoveLayerChangeset(map, layer, index);
  }

  constructor(private map: MapWrapper, private layer: LayerWrapper, private index: number) {
    super();
  }

  public async execute(): Promise<void> {
    this.map.removeLayer(this.layer);

    // We activate the last layer
    const layers = this.map.getLayers();
    if (layers.length) {
      this.map.setActiveLayer(layers[layers.length - 1]);
    }
  }

  public async undo(): Promise<void> {
    this.map.addLayer(this.layer, this.index);
    this.map.setActiveLayer(this.layer);
  }
}
