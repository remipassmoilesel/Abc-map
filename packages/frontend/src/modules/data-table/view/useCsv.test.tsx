/**
 * Copyright © 2022 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { abcRender } from '../../../core/utils/test/abcRender';
import { useCsv } from './useCsv';
import { useEffect, useMemo } from 'react';
import { LayerFactory } from '../../../core/geo/layers/LayerFactory';
import { FeatureWrapper } from '../../../core/geo/features/FeatureWrapper';
import { FileIO, FilesSelected, InputResultType } from '../../../core/utils/FileIO';
import { TestHelper } from '../../../core/utils/test/TestHelper';
import { CsvParser } from '../../../core/data/csv-parser/CsvParser';
import { VectorLayerWrapper } from '../../../core/geo/layers/LayerWrapper';
import { act } from '@testing-library/react';
import { errorMessage } from '@abc-map/shared';

jest.mock('../../../core/utils/FileIO');

describe('useCsv', () => {
  const sampleLayer = () => {
    const layer = LayerFactory.newVectorLayer();
    layer
      .getSource()
      .addFeatures([
        FeatureWrapper.create().setId('ysZrLxikgJ').setProperties({ name: 'Montpellier', id: 11 }).unwrap(),
        FeatureWrapper.create().setId('FkgKCaSCFx').setProperties({ name: 'Cherbourg', id: 22 }).unwrap(),
        FeatureWrapper.create().setId('X_-J5_VShM').setProperties({ name: 'Grenoble', id: 33 }).unwrap(),
        FeatureWrapper.create().setId('-_HQZ2Mzmj').setProperties({ name: 'Toulon', id: 44 }).unwrap(),
      ]);
    return layer;
  };

  const sampleLayerRows = (layer: VectorLayerWrapper) =>
    layer
      .getSource()
      .getFeatures()
      .map((feat) => FeatureWrapper.from(feat).toDataRow());

  it('exportFile()', async () => {
    // Prepare
    function TestComponent() {
      const layer = useMemo(() => sampleLayer(), []);
      const rows = useMemo(() => sampleLayerRows(layer), [layer]);

      const { exportFile, result } = useCsv(layer, rows);
      useEffect(() => exportFile(), [exportFile]);

      return <>{JSON.stringify({ result })}</>;
    }

    // Act
    const { container } = abcRender(<TestComponent />);
    await TestHelper.wait(10); // We wait for an internal promise

    // Assert
    expect(container).toMatchSnapshot();
    expect(FileIO.downloadBlob).toBeCalledTimes(1);

    const file = (FileIO.downloadBlob as jest.MockedFn<any>).mock.calls[0][0] as File;
    expect(file).toBeDefined();

    const rows = await CsvParser.parse(file);
    expect(rows).toMatchSnapshot();
  });

  it('importFile() should fail if content is invalid', async () => {
    // Prepare
    const prompt: FilesSelected = {
      type: InputResultType.Confirmed,
      files: [new Blob(['jsjsajsjasjajsajsjasjajs']) as File],
    };
    (FileIO.openPrompt as jest.MockedFn<any>).mockResolvedValue(prompt);

    function TestComponent() {
      const layer = useMemo(() => sampleLayer(), []);
      const rows = useMemo(() => sampleLayerRows(layer), [layer]);

      const { importFile, result } = useCsv(layer, rows);
      useEffect(() => importFile(), [importFile]);

      return <>{JSON.stringify({ result, error: errorMessage(result?.error) })}</>;
    }

    // Act
    const { container } = abcRender(<TestComponent />);
    await act(async () => {
      await TestHelper.wait(10); // We wait for an internal promise
    });

    // Assert
    expect(container).toMatchSnapshot();
  });

  it('importFile() should update layer', async () => {
    // Prepare
    const csvFile = await CsvParser.unparse(
      [
        {
          __id__: 'ysZrLxikgJ',
          name: 'Montpellier',
          id: '11',
        },
        {
          __id__: 'FkgKCaSCFx',
          name: 'Cherbourg',
          id: '22',
        },
        {
          __id__: 'X_-J5_VShM',
          name: 'Gre-gre-noble',
          id: '33',
        },
        // "Toulon" element will not be deleted because delete is not implemented.
        // This element will be skipped because we can not create a geometry for the moment.
        {
          __id__: 'C5VurN_4k_',
          name: 'Brest',
          id: '55',
        },
      ],
      'file.csv'
    );
    const prompt: FilesSelected = {
      type: InputResultType.Confirmed,
      files: [csvFile],
    };
    (FileIO.openPrompt as jest.MockedFn<any>).mockResolvedValue(prompt);

    const layer = sampleLayer();

    function TestComponent({ layer }: { layer: VectorLayerWrapper }) {
      const rows = useMemo(() => sampleLayerRows(layer), [layer]);

      const { importFile, result } = useCsv(layer, rows);
      useEffect(() => importFile(), [importFile]);

      return <>{JSON.stringify({ result })}</>;
    }

    // Act
    const { container } = abcRender(<TestComponent layer={layer} />);
    await act(async () => {
      await TestHelper.wait(10); // We wait for an internal promise
    });

    // Assert
    expect(container).toMatchSnapshot();

    const dataRows = layer
      .getSource()
      .getFeatures()
      .map((f) => FeatureWrapper.from(f).toDataRow());
    expect(dataRows).toMatchSnapshot();
  });
});
