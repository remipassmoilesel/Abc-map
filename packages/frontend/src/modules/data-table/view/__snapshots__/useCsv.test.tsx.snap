// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useCsv exportFile() 1`] = `
<div>
  {}
</div>
`;

exports[`useCsv exportFile() 2`] = `
Array [
  Object {
    "__id__": "ysZrLxikgJ",
    "id": "11",
    "name": "Montpellier",
  },
  Object {
    "__id__": "FkgKCaSCFx",
    "id": "22",
    "name": "Cherbourg",
  },
  Object {
    "__id__": "X_-J5_VShM",
    "id": "33",
    "name": "Grenoble",
  },
  Object {
    "__id__": "-_HQZ2Mzmj",
    "id": "44",
    "name": "Toulon",
  },
]
`;

exports[`useCsv importFile() should fail if content is invalid 1`] = `
<div>
  {"result":{"file":{},"imported":0,"skipped":0,"skippedIds":[],"error":{}},"error":"Invalid CSV file, nothing was parsed"}
</div>
`;

exports[`useCsv importFile() should update layer 1`] = `
<div>
  {"result":{"file":{},"imported":3,"skipped":1,"skippedIds":["C5VurN_4k_"],"error":null}}
</div>
`;

exports[`useCsv importFile() should update layer 2`] = `
Array [
  Object {
    "data": Object {
      "id": "11",
      "name": "Montpellier",
    },
    "id": "ysZrLxikgJ",
    "metadata": Object {
      "selected": false,
    },
  },
  Object {
    "data": Object {
      "id": "22",
      "name": "Cherbourg",
    },
    "id": "FkgKCaSCFx",
    "metadata": Object {
      "selected": false,
    },
  },
  Object {
    "data": Object {
      "id": "33",
      "name": "Gre-gre-noble",
    },
    "id": "X_-J5_VShM",
    "metadata": Object {
      "selected": false,
    },
  },
  Object {
    "data": Object {
      "id": 44,
      "name": "Toulon",
    },
    "id": "-_HQZ2Mzmj",
    "metadata": Object {
      "selected": false,
    },
  },
]
`;
